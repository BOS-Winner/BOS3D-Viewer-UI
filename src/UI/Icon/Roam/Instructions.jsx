import React from "react";
import ReactDOM from "react-dom";
import PropTypes from "prop-types";
import freeRoamCtrlImg from "./img/freeRoamCtrl.png";
import style from "./style.less";
// import direction from "./img/direction.png";
// import wasd from "./img/wasd.png";
// import q from "./img/q.png";
// import e from "./img/e.png";
// import mouse from "./img/mouse.jpeg";
// import plus from "./img/plus.png";
// import minus from "./img/minus.png";

class Instructions extends React.PureComponent {
  render() {
    if (this.props.show) {
      return ReactDOM.createPortal(
        <div
          className={style.instruction}
          role="presentation"
          onClick={ev => {
            ev.preventDefault();
            ev.stopPropagation();
          }}
        >
          <div className={style.title}>
            <span
              role="button"
              tabIndex={0}
              onClick={ev => {
                ev.preventDefault();
                ev.stopPropagation();
                this.props.onClose();
              }}
            >
              x
            </span>
          </div>
          <div className={style.container}>
            <div className={style.row1}>
              <div>速度</div>
              <div>视角</div>
            </div>
            <img alt="tips" src={freeRoamCtrlImg} />
            <div className={style.row3}>
              按Shift键加速
            </div>
          </div>
        </div>,
        this.props.viewportDiv
      );
    } else {
      return '';
    }
  }
}

Instructions.propTypes = {
  show: PropTypes.bool.isRequired,
  onClose: PropTypes.func.isRequired,
  viewportDiv: PropTypes.object.isRequired, // wtf!!!
};

export default Instructions;
