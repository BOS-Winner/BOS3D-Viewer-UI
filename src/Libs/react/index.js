import React from "react";
import useUpdateEffect from "./useUpdateEffect";

React.useUpdateEffect = useUpdateEffect;

export default React;
// export { default as useUpdateEffect } from "./useUpdateEffect";
