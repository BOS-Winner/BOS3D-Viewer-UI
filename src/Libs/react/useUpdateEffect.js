import { useEffect, useRef } from "react";

export default function (effect, deps) {
  const isFirstRender = useRef(true);
  useEffect(() => {
    if (isFirstRender.current) {
      isFirstRender.current = false;
    } else {
      effect();
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, deps);
}
