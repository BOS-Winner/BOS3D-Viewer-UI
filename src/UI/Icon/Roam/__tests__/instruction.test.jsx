import React from 'react';
import { render } from "@testing-library/react";
import Instructions from '../Instructions';

let getEleByRole;
let rerenderIns;
const onClose = jest.fn();

describe('instruction', () => {
  beforeEach(() => {
    const div = document.createElement('div');
    div.id = 'test';
    document.body.appendChild(div);
    const { getByRole, rerender } = render((
      <Instructions
        show
        onClose={onClose}
        viewportDiv={div}
      />
    ));
    getEleByRole = getByRole;
    rerenderIns = rerender;
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should render dom', () => {
    expect(document.body).toMatchSnapshot();
    rerenderIns((
      <Instructions
        show={false}
        onClose={onClose}
        viewportDiv={document.querySelector('#test')}
      />
    ));
    expect(document.body).toMatchSnapshot();
  });

  it('should call onClose', () => {
    getEleByRole('button').click();
    expect(onClose).toBeCalledTimes(1);
  });
});
