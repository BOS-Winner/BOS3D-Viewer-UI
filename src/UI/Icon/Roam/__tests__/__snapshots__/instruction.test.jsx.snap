// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`instruction should render dom 1`] = `
<body>
  <div
    id="test"
  >
    <div
      class="instruction"
      role="presentation"
    >
      <div
        class="title"
      >
        <span
          role="button"
          tabindex="0"
        >
          x
        </span>
      </div>
      <div
        class="container"
      >
        <div
          class="row1"
        >
          <div>
            速度
          </div>
          <div>
            视角
          </div>
        </div>
        <img
          alt="tips"
          src="freeRoamCtrl.png"
        />
        <div
          class="row3"
        >
          按Shift键加速
        </div>
      </div>
    </div>
  </div>
  <div />
</body>
`;

exports[`instruction should render dom 2`] = `
<body>
  <div
    id="test"
  />
  <div>
    
  </div>
</body>
`;
